### ISSUE TITLE
GenericStorage not creating InvestmentFlows when used with invest_relation_input_output

### ISSUE THREAD

**Issue by lensum** [oemof/oemof-solph#1013]
When using the GenericStorage with an investment optimization approach like this:
```
electric_storage = cmp.GenericStorage(
            label="electric_storage",
            inputs={el_bus: solph.Flow()},
            outputs={el_bus: solph.Flow()},
            loss_rate=0.01,  # based on 25kW for 2.5MW self consumption
            nominal_storage_capacity=solph.Investment(
                ep_costs=elstor_epc, existing=1200, maximum=2100
            ),
            invest_relation_input_capacity=0.9375,
            invest_relation_input_output=1,
            inflow_conversion_factor=0.9,
            outflow_conversion_factor=0.8 / 0.9,  # overall efficiency of 0.8
            initial_storage_level=0,
            balanced=False,
            ),
        )
```
I get `KeyError: 'Index \'("<oemof.solph.components._generic_storage.GenericStorage: \'electric_storage\'>", "<oemof.solph.buses._bus.Bus: \'el_bus\'>", 0)\' is not valid for indexed component \'InvestmentFlowBlock.total\''`

I assume that the error might be in the internal function `_set_flows`. I defined the `self.invest_relation_input_capacity` and `self.invest_relation_input_output`, but the `_set_flows` function only considers `self.invest_relation_output_capacity` for the output flows, meaning that my output flows don't get converted into `InvestmentFlow` objects.
If I use `invest_relation_output_capacity=0.9375` instead of the `invest_relation_input_output`, the model starts to solve.

**Comment by p-snft** [comment]
Shouldn't the `Flow`s be created with `nominal_value=Investment()` to allow for investments in any case?

**Comment by p-snft** [comment]
Maybe, it makes sense to rewrite this thing anyway. I'd consider replacing the fixed relations between the investments by a limit to the c-rate. It would be an internal property that exists independent from the nominal values of the flows.

**Comment by jokochems** [comment]
Yeah, you are right about locating this to the `_set_flows` method, @lensum , I think.
* The case of `invest_relation_input_output` is not accounted for here. In my view, a check needs to be added in `_set_flows`.
* Also, for the other cases, the type of `flow.investment` is evaluated. I think, this needs to be `flow.nominal_value` instead according to the latest structural changes.

**Comment by p-snft** [comment]
I see multiple problems with the current implementation. First of all, we create unexpected investment variables. (If the Flows do not have to have one, they are silently created.) Secondly, you can do confusing things. The `invest_relation_input_capacity=0.1` will be typically read as "it needs 10 hours to charge the storage". However, you can set a "max" to the flows to bypass that:

```python
electric_storage = cmp.GenericStorage(
    label="electric_storage",
    inputs={el_bus: solph.Flow(max=10)},  # will be charged in 1 hour
    outputs={el_bus: solph.Flow()},
    nominal_storage_capacity=solph.Investment(
        ep_costs=elstor_epc, existing=1200, maximum=2100
    ),
    invest_relation_input_capacity=0.1,
    invest_relation_input_output=0.1,
)
```

**Comment by lensum** [comment]
I'm currently working on a bugfix for this, but encounter some hurdles while installing a solver (so it might take a little time for me to open the official PR).
The bugfix will not include changes to the (expected) behavior. I think it would be best to discuss and then maybe implement the changes you proposed @p-snft when we revise the GenericStorage.

**Comment by lensum** [comment]
Hey @p-snft,
while working on the bugfix, I actually decided to follow your suggestion to put an end to the implicit generation of `Investment` objects. It makes the behaviour much more "deterministic" (i.e. easier to understand) and also reduces the amount of work in testing. Plus, I think it is actually a good thing to make the users think about which `Investment` objects they need and what their parameters should be.

I will try to work on the remaining pipeline error(s) tomorrow, other than that the PR is my suggestion on how to handle this problem. What I mean: functionality-wise, the PR should be complete.

**Comment by p-snft** [comment]
Has been fixed in #1217.
